### ISSUE TITLE
async fn returning boxed dyn trait requires explicit coercion

### ISSUE THREAD

**Issue by octavore** [rust-lang/rust-analyzer#21477]
**rust-analyzer version**: `rust-analyzer version: 0.3.2753-standalone (d43c636243 2026-01-12) [/Users/*/vscode/extensions/rust-lang.rust-analyzer-0.3.2753-darwin-arm64/server/rust-analyzer]`

**rustc version**: `rustc 1.91.1 (ed61e7d7e 2025-11-07)`

**editor or extension**: VSCode, 0.3.2753

**code snippet to reproduce**:

The following compiles but produces an error in VSCode. The non-async version of the function does not produce the same error.

```rust

trait Credential {}

struct UserCredential(i32);

impl Credential for UserCredential {}

struct OtherCredential {}

impl Credential for OtherCredential {}

struct Session {
    user_id: Option<i32>,
}

async fn get_credential(session: &Session) -> Option<Box<dyn Credential>> {
    if let Some(user_id) = session.user_id {
        return Some(Box::new(UserCredential(user_id)));
    }
    // error in vscode: expected UserCredential, found OtherCredential, E0308
    Some(Box::new(OtherCredential {}))
}
```

Adding explicit coercion to `Box<dyn Credential>` or `Box<_>` does fix the rust analyzer error

```rust
async fn get_credential(session: &Session) -> Option<Box<dyn Credential>> {
    if let Some(user_id) = session.user_id {
        return Some(Box::new(UserCredential(user_id)) as Box<_>);
    }
    Some(Box::new(OtherCredential {}) as Box<_>)
}
```

**Comment by ChayimFriedman2** [comment]
That's probably something with how we handle desugaring of async fns causing it to not be a coercion site.

**Comment by ChayimFriedman2** [comment]
The reason for this bug is mishandling of `CoerceMany` in async blocks. The real fix is to migrate from a special `Expr::Async` to be like rustc, where async blocks are a kind of closure. I'm currently mid-way refactoring the entirety of call and closure analysis in inference, I'll do that after.

**Comment by kext** [comment]
I ran into this when returning an empty slice. I suppose this is the same bug?
```rust
async fn foo(a: &[i32], n: usize) -> &[i32] {
    if n > a.len() {
        return &[];
    }
    // error in vscode: expected &[{unknown}; 0], found &[i32]
    &a[..n]
}
```
This version on the other hand is fine, I guess because the complete if is only one return.
```rust
async fn foo(a: &[i32], n: usize) -> &[i32] {
    if n > a.len() {
        &[]
    } else {
        &a[..n]
    }
}
```
